packages/core/src/interpreter.ts

```diff
@@ -404,48 +404,39 @@ export class Interpreter<
       | ((
           state: State<TContext, TEvent, any, TTypestate, TResolvedTypesMeta>
         ) => void)
-      | Observer<State<TContext, TEvent, any, TTypestate, TResolvedTypesMeta>>,
+      | Partial<
+          Observer<State<TContext, TEvent, any, TTypestate, TResolvedTypesMeta>>
+        >,
     _?: (error: any) => void, // TODO: error listener
     completeListener?: () => void
   ): Subscription {
-    if (!nextListenerOrObserver) {
-      return { unsubscribe: () => void 0 };
-    }
-
-    let listener: (
-      state: State<TContext, TEvent, any, TTypestate, TResolvedTypesMeta>
-    ) => void;
-    let resolvedCompleteListener = completeListener;
-
-    if (typeof nextListenerOrObserver === 'function') {
-      listener = nextListenerOrObserver;
-    } else {
-      listener = nextListenerOrObserver.next.bind(nextListenerOrObserver);
-      resolvedCompleteListener = nextListenerOrObserver.complete.bind(
-        nextListenerOrObserver
-      );
-    }
+    const observer = toObserver(nextListenerOrObserver, _, completeListener);
 
-    this.listeners.add(listener);
+    this.listeners.add(observer.next);
 
     // Send current state to listener
     if (this.status !== InterpreterStatus.NotStarted) {
-      listener(this.state);
+      observer.next(this.state);
     }
 
-    if (resolvedCompleteListener) {
-      if (this.status === InterpreterStatus.Stopped) {
-        resolvedCompleteListener();
-      } else {
-        this.onDone(resolvedCompleteListener);
-      }
+    const completeOnce = () => {
+      this.doneListeners.delete(completeOnce);
+      this.stopListeners.delete(completeOnce);
+      observer.complete();
+    };
+
+    if (this.status === InterpreterStatus.Stopped) {
+      observer.complete();
+    } else {
+      this.onDone(completeOnce);
+      this.onStop(completeOnce);
```

Comment:
this was a little bit annoying, but I think it's the cleanest solution possible right now. This probably will be completely refactored in v5 anyway.